tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc

```diff
@@ -2998,6 +2999,138 @@ LogicalResult ConvertTFLCosOp::matchAndRewrite(
   return success();
 }
 
+LogicalResult ConvertTFLAtan2Op::matchAndRewrite(
+    Operation* op, PatternRewriter& rewriter) const {
+  auto tfl_atan2_op = cast<TFL::Atan2Op>(op);
+  Location loc = op->getLoc();
+  Value input_y = tfl_atan2_op.getY();
+  RankedTensorType input_y_ty = input_y.getType().dyn_cast<RankedTensorType>();
+  Value input_x = tfl_atan2_op.getX();
+  RankedTensorType input_x_ty = input_x.getType().dyn_cast<RankedTensorType>();
+  ShapedType output_ty =
+      tfl_atan2_op.getResult().getType().dyn_cast<ShapedType>();
+
+  if (!input_y_ty || !input_x_ty) {
+    return rewriter.notifyMatchFailure(
+        op, "ConvertTFLAtan2Op: ranked inputs required");
+  }
+
+  Type input_y_ety = input_y_ty.getElementType();
+  Type input_x_ety = input_x_ty.getElementType();
+  Type output_ety = output_ty.getElementType();
+
+  bool op_is_fp = input_y_ty.getElementType().isF32();
+  if (!op_is_fp) {
+    return rewriter.notifyMatchFailure(
+        op, "ConvertTFLAtan2Op: input/result must be fp32.");
+  }
+
+  // To perform an atan2 operation we make use of an atan lookup table,
+  // then determine the correct quadrant for each output. To restrict the
+  // input domain of the lookup table from [-inf, inf] to [0, 1], we make
+  // use of two identities and undo the transformation later on:
+  //
+  // acrtan(z) = π/2 - arctan(1/z)                                  (0)
+  //
+  // and
+  //
+  // arctan(-z) = -arctan(z)                                        (1)
+
+  Value pi = getTosaConstTensorSingleF32(rewriter, op, M_PI);
+  Value pi_2 = getTosaConstTensorSingleF32(rewriter, op, M_PI_2);
+  Value zero = getTosaConstTensorSingleF32(rewriter, op, 0.0);
+  Value one = getTosaConstTensorSingleF32(rewriter, op, 1.0);
+  Value two = getTosaConstTensorSingleF32(rewriter, op, 2.0);
+
+  // 1. Restrict the input to the atan lookup from [-inf, inf] to [0, 1].
+  // By utilizing (0) and (1) we compute: min(|x|, |y|) / max(|x|, |y|).
```

Comment:
Do we need to handle case where |x| and |y| are both 0?

Reply:
For some reason I don't see your responses.

Reply:
Having consistent behavior is nice, but this is undefined behavior, so it may be impossible to guarantee consistency in all cases.

This sequence of ops does have defined behavior in TOSA, according to the current spec. We don't define reciprocal(0) explicitly, but we do say non-zero / 0 = infinity, and then 0 * infinity = 0.

Based on this, I think this solution looks good.